azavea/docker-openjdk-gdal
# Create Dockerfiles for Slim and Alpine builds

## Overview

Define two Dockerfile templates for Slim and Alpine builds, each of which compiles GDAL from source using Java and Python bindings. Bootstrap the Dockerfile templates off of https://github.com/geotrellis/geotrellis-gdal, with minor modifications to make builds faster.

Closes https://github.com/geotrellis/geotrellis-server/issues/60.

## Notes

* We had debated switching to an approach that would use the prebuilt [`libgdal-java`](https://packages.debian.org/stretch/libgdal-java) package to install these bindings. While this approach would benefit from faster build times and less maintenance on our end, I moved away from it for a few reasons:
    * `libgdal-java` is not available on `apk`
    * Nathan and Grigory both mentioned wanting to use the latest version of Geotrellis, which is only available on `experimental` Debian builds
    * After an hour or so of tweaking I couldn't get the package installation to pass the tests in https://github.com/geotrellis/geotrellis-server using a `sid-slim` build

## Testing instructions

* Confirm that Slim and Alpine builds compile successfully on Travis
* Run `GDAL_VERSION=2.3.2 OPENJPEG_VERSION=2.3.0 VARIANT=slim ./scripts/cibuild` to build a Slim image (note that this can take ~30 minutes, so plan ahead)
* Clone https://github.com/geotrellis/geotrellis-server and `cd` into it
* Run the following commands to test the images against real Geotrellis tests:

```
docker run \
  -v $(pwd)/build.sbt:/root/build.sbt \
  -v $(pwd)/project:/root/project \
  -v $(pwd)/core:/root/core \
  -v $(pwd)/sbt:/root/sbt \
  -w /root \
  quay.io/azavea/openjdk-gdal:2.3.2-slim \
  ./sbt "project core" test
```

* Clone https://github.com/geotrellis/geotrellis-gdal/ and `cd` into it
* Run the following commands to test the images against more real Geotrellis tests:

```
docker run \
  -v $(pwd)/build.sbt:/root/build.sbt \
  -v $(pwd)/project:/root/project \
  -v $(pwd)/gdal:/root/gdal \
  -v $(pwd)/src:/root/src \
  -v $(pwd)/sbt:/root/sbt \
  -w /root \
  quay.io/azavea/openjdk-gdal:2.3.2-slim \
  ./sbt "project gdal" test
```
